### ISSUE TITLE
[RFC] Possible handling of toolchains in `alr`

### ISSUE THREAD

**Issue by mosteo** [alire-project/alire#625]
A few open questions about toolchain handling.

- Should Alire install toolchains?
   - From distrib? (msys2, debian, etc.)
   - From GNAT Community?
- Should Alire have its own toolchain builds?
- Cross toolchains?

**Comment by wiremoons** [comment]
Hi @mosteo 

**TL;DR:** : I have been thinking about the suggestions / questions you have made for this RFC. The ability for `alr` to bootstrap an Ada GNAT tool chain install if it is missing on a system would be a good feature in my view! 

For more thoughts if you have time to read them:

I have used the similar capability available with the OCaml `opam` tool which can do this already. As you may know already, basically, by just using the single `opam` pre-built binary you can bootstrap and install the OCaml compiler , tools, and libraries (at least on Linux anyway) which is great to manage the whole process with one application, and get an OCaml development environment up and running, without needing much knowledge about such complexities!

Currently I have automated the install of AdaCore GNAT Community Edition 2020 for Windows 10 using a PowerShell script. I have also automated the same functionality on Ubuntu 20.04 (via a bash shell script) to either do the same (ie install of GNAT Community Edition 2020) or install the local package manager version of GNAT 9.x using `apt`.  

For systems that have local package manager install options, then it would be good to give the user a choice - otherwise, if they have no opinion either way, the AdaCore GNAT Community Edition 2020 could be provided, as that is probably an easier starting state for many... I accept other users would beg to differ perhaps :) But this is about bootstrapping to allow someone to get up an running and developing in Ada, or using an opensource tools they have found, or maybe wanting to try IoT development tutorials etc. The last thing a new user needs is to get 'lost' in tool chains, libraries, paths, build issues and so on! They can learn that later as they become more advanced and/or interested. 

If `alr` could include this capability it would be useful, and save maintaining different processes to set a computer up. The use of `alr` to enable this, and kick off using a computer to develop and build Ada applications would be beneficial in many ways!

Do you have any additional thoughts on:

1.) Would this be integrated into `alr` directly? Or would it be an additional tool / package available via 'Alire' instead? If separate it would need to be downloadable by `alr` as a pre-complied static binary of course too, other wise it would be 'chicken and egg' again, as no Ada GNAT tool chain is currently available on the system concerned to actually build the Alire package :) The approach of separating it out into a different application would help off load complexity from `alr` too. Not sure if it would be worth it though..? As long as the process was seamless to a user - I don't suppose it matters much?

2.) As mentioned above - I think the system `alr` is running on should be detected (which it does already!), and depending on the result either a local package manager install could be automated (again assuring a working tool chain and basic libraries to get going!) - or offer GNAT Community also - as some may prefer it. For systems like Windows - then just installing GNAT Community would be a great starting position in my view.

3.) An alternative would be to go all in on local package managers per platform. Easier for Linux based system than others. However Windows 10 has a few too now of course (`scoop`; `chocolatey`; etc, and newer opensource `winget-cli` from Microsoft). And I believe macOS has `brew` - but I havent used a Apple computer in a while, so not sure how well it supports Ada tool chains already, especially as it defaults to CLANG/LLVM with XCode not GCC... FreeBSD and other BSD are probably in a similar position... Perhaps a focus on ensuring packages are available through these routes would be more inline with existing system practices? The downside in my view is that a user would need to know how to use the packages manager, and what packages to install sometimes too. Plus the outside dependency to ensure the packages are all kept up to date and offer functionality that works well with `alr`.  A way to help create a 'standard' in this scenario would be to provide universal opensource GNAT with suitable licensing across all these platforms and package management tools - a task for one open source project (ie Alire) would be a bit extreme to take on...  So - circular argument maybe - and back to using `alr` to bootstrap a Ada toolchain for the majority would be much easier in many ways! It could just use the already available options - but manage and automate the process for the user. 

 3.) I think your 'cross tool chain' question can be answered more easily once an approach is agreed? I do think an option to also provide it (when available) would be good - for many of the same reasons outlined above (accessibility, least path or resistance for a new user, open Ada IoT development to more people; etc). 

4.) The Windows 10 install and provision of msys2 by `alr` works very well in my experience, and is a great feature. The fact this is installed (and similar tools can easily be installed via local package managers on other systems too) may allow Ada GNAT to be built (if needed)? Not sure how long this would take on an average persons computer to build up a working GNAT tool chain from source?  Can the distribution of GNAT be merged into something like msys2 instead?  Not sure why this would be a sensible undertaking for `alr` directly, as using existing pre-built downloadable tool chains such as GNAT Community would be much easier! So  just a thought - as I am not expert in any of this :)

Anyway - enough rambling from me! 

Hopefully others will have constructive and better informed opinions on how this could be achieved using `alr`. As stated - personally I thing the ability for `alr` to bootstrap a GNAT tool chain would be great.

**Comment by Fabien-Chouteau** [comment]
Thank you @wiremoons for the feedback :+1: 

> 1.) Would this be integrated into `alr` directly?

Yes, I don't see the point of having a separate tool for this. Also because one could specify dependency on a given version of the toolchain so it fits in Alire features. Everything we need to download an install a toolchain archive is more or less already in Alire.

> 2.) As mentioned above - I think the system `alr` is running on should be detected (which it does already!), and depending on the result either a local package manager install could be automated (again assuring a working tool chain and basic libraries to get going!).

This could be an option indeed, I am worried that it would be too complicated for beginners.
 
> 3.) An alternative would be to go all in on local package managers per platform. Easier for Linux based system than others.

Actually Linux might be more diffucult. For Windows we have a solution that works, msys2, and it work for "any" Windows setup.
However for Linux, not all distributions have a gprbuild and GNAT package, and none of them have the cross versions (arm, RISC-V, etc.).

> Perhaps a focus on ensuring packages are available through these routes would be more inline with existing system practices? 

It's not clear if we will be able to achieve that in a short term. I started to add stuff in msys2 for Windows, and I asked for a cross compiler on Arch Linux (https://bugs.archlinux.org/task/69383). But there is not enough traction in the Ada community at the moment to get all the toolchains on all the distribs. At least that's my opinion.

> 4.) The Windows 10 install and provision of msys2 by `alr` works very well in my experience, and is a great feature.

Thank you, I worked hard to get this in Alire :)

> The fact this is installed (and similar tools can easily be installed via local package managers on other systems too) may allow Ada GNAT to be built (if needed)? Not sure how long this would take on an average persons computer to build up a working GNAT tool chain from source?

Building GCC and GNAT can take a huge amount of time, for the arm-elf cross compiler it takes more than 5 hours (on a beefy machine). 

> Can the distribution of GNAT be merged into something like msys2 instead?

For Windows that's what I am aiming for. Then there is the question of availability of versions since msys2 only keeps the latest version of packages. One option for me is to repackage the msys2 binaries and keep them in a repo so Alire can get even the old versions.

> Not sure why this would be a sensible undertaking for `alr` directly, as using existing pre-built downloadable tool chains such as GNAT Community would be much easier!

The GNAT Community release might stop in the future: https://www.reddit.com/r/ada/comments/hwgbwa/survey_on_the_future_of_gnat_community/

**Comment by mosteo** [comment]
Hi @wiremoons, thanks for the feedback! I don't have much to add, but personally I would like to see it happen, specially if CE editions disappear in the future.

Fabien's point on compilation time is important, but I'd still would like to have it as a fallback, if it is not outrageously difficult to package. Other than that, I feel that distro-packaged versions should be the default whenever available.

Right now we already use dependencies on `gnat` to signal particular versions that are known to [not] work. I don't think we should offer to install a huge toolchain just to satisfy a dependency, but once we have `alr install` (#482), that could be the way to provide a default user-wide compiler, and even workspace-specific versions.

**Comment by shiz01** [comment]
Yes, it's very convenient. I used this functionality in rustup, a tool for installing the rust toolchain.
Distribution format? I think the prebuild toolchain will be the most convenient for new users. Then can add an assembly from the source code.

Add support installation in a local directory without root privileges:
`${HOME}/.alr/toolchain/gnat-${ARCH}-${KERNEL}-${OS}/${version}`
`${HOME}/.alr/toolchain/gnat-x86_64-linux-gnu/11.1.0`
`${HOME}/.alr/toolchain/gnat-x86_64-linux-musl/9.3.0`


Something like this would be handy:

`alr toolchain list`
_gnat-aarch64-linux-gnu
gnat-riscv32-none-elf
gnat-riscv64-none-elf
gnat-x86_64-linux-gnu
...
spark-x86_64-linux-gnu
..._

`alr toolchain info gnat-riscv64-none-elf`
Available hosts:
linux - aarch64, x86_64, i386
Windows - x64 ,i386
MacOS - x64 ,i386
Available versions:
8.4.0
10.3.0
11.1.0
Installled versions: 8.4.0 11.1.0
Home page: link to gcc.gnu.org
Description: The GNU Compiler Collection, for riscv64

By default auto detect host:
`alr toolchain install gnat`

Define host manualy:
`alr toolchain install gnat --host x86_64-linux-gnu`

Full specified name:
`alr toolchain install gnat-x86_64-linux-gnu`
`alr toolchain install gnat --host x86_64-linux-gnu`

Specify version:
`alr toolchain install gnat --version 9.3.0`
`alr toolchain install gnat --version 11.1.0`
`alr toolchain install spark -v latest`

Set current version:
`alr toolchain list --installed`

Check current default compiller:
`alr toolchain default`
_gnat x86_64-linux-gnu 11.1.0
spark x86_64-linux-gnu 2019_

Set default toolchain:
`alr toolchain default gnat -v 9.3.0`
`alr toolchain default spark -v 2021`

Remove all version gnat:
`alr toolchain remove gnat`

Remove specific version:
`alr toolchain remove gnat -v 9.3.0`
`alr toolchain remove gnat-9.3.0`

Cross Compile:
`alr toolchain install gnat-arm-none-eabi --host x86_64-linux-gnu`
`alr toolchain install gnat --target arm-none-eabi --host x86_64-linux-gnu`

**Comment by Fabien-Chouteau** [comment]
@shiz01 right now the plan is to have the toolchains almost exactly like regular crates. Which means there won't be any special commands to declare a dependency and install the toolchain. For example:
```console
$ alr init test --bin --in-place
$ alr with hal # Add a dependency to a library
$ alr with gnat_arm_elf # Add a dependency to a cross toolchain
```

**Comment by mosteo** [comment]
I think most if not all that was discussed here was already merged in #787 and follow-ups. If there is some specific feature that isn't satisfied please open a new focused PR.
